server/rooms.ts

```diff
@@ -1363,17 +1366,53 @@ export class GlobalRoomState {
 			// kill server in 10 seconds if it's still set to
 			setTimeout(() => {
 				if (Config.autolockdown && Rooms.global.lockdown === true) {
-					// finally kill the server
-					process.exit();
+					// finally save state / kill the server
+					void this.kill();
 				} else {
 					this.notifyRooms(
 						notifyPlaces,
-						`|html|<div class="broadcsat-red"><b>Automatic server lockdown kill canceled.</b><br /><br />In the last final seconds, the automatic lockdown was manually disabled.</div>`
+						`|html|<div class="broadcast-red"><b>Automatic server lockdown kill canceled.</b><br /><br />In the last final seconds, the automatic lockdown was manually disabled.</div>`
 					);
 				}
 			}, 10 * 1000);
 		}
 	}
+	async kill() {
+		if (Config.persistbattles) await this.writeBattleState();
+		return process.exit();
+	}
+	async writeBattleState() {
+		const buffer: AnyObject = {};
+		for (const [id, room] of Rooms.rooms) {
+			if (!room.battle) continue;
+			const formatid = room.battle.format;
+			const logData = await room.battle.getLog();
+			if (!logData) continue;
+			if (!buffer[formatid]) buffer[formatid] = {};
+			buffer[formatid][id] = {
+				inputLog: logData.map(item => item.replace(/\r/g, '')).join('\n'),
+				title: room.title,
+				num: room.roomid.split('-')[2],
+			};
+		}
+		return FS(`logs/battles.json`).writeUpdate(() => JSON.stringify(buffer));
+	}
+	loadBattleState() {
+		if (!Config.persistbattles) return;
+		const battleData = JSON.parse(FS(`logs/battles.json`).readIfExistsSync() || "{}");
```

Comment:
We should probably delete battle state after reading it, or it could cause crash loops and other bad crash behavior.